support for filter_by_tag_values, for filtering on extra asg tags

tootedom opened the pull request:
# Issue Type

- Feature Pull Request


## Summary

Supporting the ability to filter on matched (found) ASG's by extra tags.
For example, we have an AWS account that contains many VPCs.  These VPCs contain different versions of the same ASGs (denoting the environment the ASG belongs to)

These different ASGs are tagged:

- Key:Environment 
   Value:dev

- Key:Environment 
   Value:qa

We'd like to run against only those ASG's tagged with "Environment" with the value "qa" for example, rather than picking up all the ASGs.  


## Code contribution checklist

Apologies for this, but I've not figured out a way in which test the changes (unit test wise).  Do you have an example anywhere about how I'd go about mocking DescribeAutoScalingGroupsPages autoscale call?

<!--
The code review should be largely a matter of going through this checklist.
-->

1. [ ] ~~The contribution fixes a single existing github issue, and it is linked
   to it.~~
1. [x] The code is as simple as possible, readable and follows the idiomatic Go
  [guidelines](https://golang.org/doc/effective_go.html).
1. [ ] All new functionality is covered by automated test cases so the overall
  test coverage doesn't decrease.
1. [x] No issues are reported when running `make full-test`.
1. [ ] ~~Functionality not applicable to all users should be configurable.~~
1. [x] Configurations should be exposed through Lambda function environment
   variables which are also passed as parameters to the
   [CloudFormation](https://github.com/cristim/autospotting/blob/master/cloudformation/stacks/AutoSpotting/template.json)
   and
   [Terraform](https://github.com/cristim/autospotting/blob/master/terraform/autospotting.tf)
   stacks defined as infrastructure code.
1. [ ] ~~Global configurations set from the infrastructure stack level should also
   support per-group overrides using tags.~~
1. [x] Tags names and expected values should be similar to the other existing
   configurations.
1. [ ] ~~Both global and tag-based configuration mechanisms should be tested and
  proven to work using log output from various test runs.~~
1. [x] The logs should be kept as clean as possible (use log levels as
  appropriate) and formatted consistently to the existing log output.
1. [x] The documentation is updated to cover the new behavior, as well as the
   new configuration options for both stack parameters and tag overrides.
1. [x] A code reviewer reproduced the problem and can confirm the code
  contribution actually resolves it.

cristim commented:
Thanks for your contribution.

I'm not sure I understand your use case, AutoSpotting won't pick all your groups, it just considers those with the `spot-enabled` tag set to `true`. Can't you just set that tag to true on the QA groups and leave it unset or set it to false everywhere else?

I guess we could also introduce this feature but use it to define the tag people use instead of hardcoding the `spot-enabled=true` tag.

@xlr-8 what are your thoughts on this?

@tootedom for the tests, have a look at our existing tests, there should be plenty such examples. For mocking we use the standard interfaces exposed by the AWS SDK for Go for mocking purposes.

tootedom commented:
>>
>> I'm not sure I understand your use case, AutoSpotting won't pick all your groups, it just 
>> considers those with the spot-enabled tag set to true. Can't you just set that tag to true on 
>> the QA groups and leave it unset or set it to false everywhere else?
>>

Apologies, I'll explain more.  We have an AWS Account.  This contains ASG for multiple environments:

- dev
- qa
- stage.

Dev is deployed to on commit from git, as a result it see's a lot of churn of instances during the day (9am to 5pm), for this the on demand per second billing is preferable (as if we use spot instances and terminate them ourselves due to new commit push, we will pay for 1 hour of use, even if we get 10 mins of runtime).  For the rest of the time ( 5pm to 10pm ), the dev environment is pretty static, with only minor updates for pushes made by dev teams in other countries.  As a result, we'd like to us autospotting on "dev", but only run it at 5pm to 10pm.

Our qa and stage environments see less flux in changes, and we can run the autospotting on the ASG's in those environments more frequently.  

As the environment have a difference in instance churn, the times we wish to run autospotting varies based on this activity.

----
>>
>> For mocking we use the standard interfaces exposed by the AWS 
>> SDK for Go for mocking purposes.
>>

Is there a specific example you can point me at for where a aws mock is being used for returning a list of matching ASGs and tags for say a "DescribeTagsInput" call.  

For example I was looking for a test that covered that "scanForEnabledAutoScalingGroups" in region.go is being executed and returning the expected list of ASGs, but I was struggling to find an example, or where to put it.  (apologies.. new to go, so I might not be looking in the correct place.  region_test.go and autospotting_test.go, don't seem to assert on the asg's returned (again, I could be looking at the incorrect thing).

xlr-8 commented:
@tootedom 
> Supporting the ability to filter on matched (found) ASG's by extra tags.

So [following your last comment ](https://github.com/cristim/autospotting/pull/209#issuecomment-361209301)in the end the question it's not about matching a specific tag but running at certain hours, is it correct?

Another point that you mentioned:

> the on demand per second billing is preferable (as if we use spot instances and terminate them ourselves due to new commit push, we will pay for 1 hour of use, even if we get 10 mins of runtime)

My understanding of their pricing, was that even spot instances were concerned by the billing per second. ref: https://aws.amazon.com/blogs/aws/new-per-second-billing-for-ec2-instances-and-ebs-volumes/ & https://aws.amazon.com/ec2/faqs/ - section Spot instances. If you understood otherwise, or if your billing shows otherwise, then please let me know.

Now, you're right in the fact that autospotting isn't that suitable for "highly versatile" ASG. Meaning it works best with low velocity. So the pricing to the second, might not change your opinion on that, but please check/consider it anyway.

If my understanding regarding your need for a time range is correct, then I do not understand your need for adding an extra tag to filter with. Indeed if you were already to play with tags on your ASGs (dev/preprod/qa/etc), then the logic would be the same but with the "spot-enabled" tag key. And adding a small script in a cron doing that, might be the easiest for you.

Please let us know if those assumption are correct, as for now it's a bit difficult to understand the need nor the benefit of such functionality.

tootedom commented:
Hi there,

>>
>> My understanding of their pricing, was that even spot instances were concerned by the
>> billing per second
>>

My bad, I'd not seen that the per second also applies to spot instances.

>>
>> And adding a small script in a cron doing that, might be the easiest for you
>>

Interesting.  I'd thought about this, but wasn't sure how I could implement it until now (for some reason).  Needed the nudge through a conversation.

I could have an additional lambda that runs on a Cloud Watch event, and adds the flag "spot-enabled=true" on the instances at 5pm, and then removes at X time. 

I agree.  I don't think the functionality is required given I can implement a CWE that can run a lamdba to add an additional tag to specific asgs.

Thanks for the help guys.  I'm happy to close the pull as "no need to implement" if you are.

cheers
/dom

cristim commented:
I agree that a Lambda with cron event should do for you, but I think it's valuable to allow people to use whatever tags they see fit, much like initially proposed, and keep the `spot-enabled=true` as merely a default value for that. I've seen the same issue raised by @chrisurwin who needed something else for his Rancher integration fork.

@tootedom if you don't mind some effort, I'd like to have this work polished until we can merge it, rather than abandon it.

I don't think we have any tests for such paginated input, but have a look at the way we define mocks in [mock_test.go](https://github.com/cristim/autospotting/blob/master/core/mock_test.go), and then we use them intensively in [autoscaling_test.go](https://github.com/cristim/autospotting/blob/master/core/autoscaling_test.go).

tootedom commented:
>>
>> So if you don't mind some effort,
>> I'd like to have this work polished until we can merge it, rather than abandon it.
>>

I'll have a look and see where I get with it this week.

cheers
/dom

chrisurwin commented:
@tootedom  https://github.com/chrisurwin/autospotting/commit/a82ef46afb1a7ba0276efc330d13fd17f543d028 shows how I did it, very simple changes, just making it an additional command line parameter and than adding it to config and updating regions to use the config rather than hard coded

tootedom commented:
Hi there,

Apologies for the delay, I've refactored it so that you can filter the ASG on as many tags as required by the user:

```
 -filter_by_tag 'spot-enabled=true' -filter_by_tag 'environment=dev' -filter_by_tag 'team=interactive'
```

The default is to filter on spot-enabled=true (as previous).   We tried autospotting on an environment with 250 ASGs today; and with 3GB of memory an 5 min run time; we could not get the lambda to complete.  This left a large number of spot instances untagged and detach from the asg.

Our hope is to execute the autospotting code, via many invoke that will allow use to filter the ASGs by "spot-enabled true' , environment, and the team that owns the ASG.  This will hope smooth out the operation of the lambda (let it finish, and use less memory).

Again, apologies for the delay in getting this back to you.  Let me know if you need/spot any required changes.

cheers
/dom

cristim commented:
Thanks for this contribution, I hope I will have time to review it later today.

Regarding the failure on large environments, I've seen it reported a lot recently. It's the next thing on my agenda but it will require a large scale refactoring and I don't know when I can get to it.

Have a look at the recent discussion on the gitter room for some implementation ideas.

tootedom commented:
I just realised I can probably get rid of an asg code.  Let me check that, and I'll get back to you.  It will reduce the logic (and remove an API request)

tootedom commented:
>>
>> Please also expose it to the Cloudformation and Terraform infrastructure code
>>

Can do.  Although I might need a pointer on this.   Would you happen to know how to specify that you can pass many 'filter_by_tag' options (not just the one) in the Cloudformation and Terraform?

cheers
/dom

cristim commented:
Lamda works with environment variables, so we have to find a way to provide these inputs in such a format.

The main feature of the flag parser library we use is that it out of the box accepts the `FOO=bar` environment variable instead of the `-foo bar` command-line flag, so these become interchangeable. 

So at a glance, to make Lambda happy we need to merge those repeated flags into a single CSV value, something like `TAG_FILTERS="spot-enabled=true,env=qa"`

To get this, the command-line parser would simply need to be refactored, converting the initially proposed
`-filter_by_tag 'spot-enabled=true' -filter_by_tag 'environment=dev' -filter_by_tag team=interactive` 
into
`-tag_filters 'spot-enabled=true,environment=dev,team=interactive'`

We used this pattern when we implemented the `-regions` flag, propagated to Lambda as the `REGIONS` environment variable. Have a look how that's done here https://github.com/cristim/autospotting/blob/7e35ee56e899bd7d240632cdc8a3fa0b61ce9a07/core/region.go#L45-L49

tootedom commented:
Cheers for that pointer.  I'll have a look later, and hopefully have an update for you.
cheers
/dom

tootedom commented:
Hi there,

I've updated as mentioned to support:

```
-tag_filters 'spot-enabled=true,environment=dev,team=interactive'
```

It also supports many appearances of -tag_filters on the command line
```
-tag_filters=spot-enabled=true -tag_filters='environment=qa,team=interactive'
```

----

I think I've update the correct references in the cloudformation and terraform for TAG_FILTERS environment variable.

let me know what you think.  I've been running this off a branch in our environment that was struggling with many asgs and instances.  Running it on different Cloudwatch Events and filtering for different ASGs has worked for us.

cheers
/dom

cristim commented:
Looks good at a first glance, thanks!
I'll check it out in more detail soon.

@xlr-8 do you have any thoughts on this?

xlr-8 commented:
@cristim I'll do my best to give it a look soon by the end of the week/next week

cristim commented:
@tootedom please resolve the merge conflicts

tootedom commented:
No problems will have a look tmr.  (Things must have moved on since the other day).
Will post at some point tmr.

tootedom commented:
Have merged in master.

xlr-8 commented:
I'm currently reviewing it!

xlr-8 commented on core/region_test.go:
```diff
@@ -291,3 +292,162 @@ func TestContainsString(t *testing.T) {
+	tests := []struct {
+		name    string
+		want    []string
+		tregion *region
```
Is it typo? Or for `test region` or a copy/paste?

tootedom replied:
Went with the consistency from the TestRequestSpotInstanceTypes(..) test.

xlr-8 commented on core/region_test.go:
```diff
@@ -291,3 +292,162 @@ func TestContainsString(t *testing.T) {
+				asgNames = append(asgNames, name.name)
+			}
+			if !reflect.DeepEqual(tt.want, asgNames) {
+				t.Errorf("region.requestSpotInstanceTypes() = %v, want %v", asgNames, tt.want)
```
I assume you meant `region.scanForEnabledAutoScalingGroups()` ;)

tootedom replied:
yup. sorry.

xlr-8 commented on terraform/autospotting/variables.tf:
```diff
@@ -46,6 +46,10 @@ variable "autospotting_regions_enabled" {
 
+variable "autospotting_tag_filters" {
+  description = "tags to filter which ASGs autospotting considers"
+}
```
I would improve a bit the description (both here and in the module) to let the user know it supports both comma & space separation, and that is a list. A bit like for asg_disallowed_instance_types, I would also add the default value to spot-enable=true
Like you did in Cloudformation :)

xlr-8 commented on core/config.go:
```diff
@@ -35,4 +77,8 @@ type Config struct {
+
+	// Filter on extra ASG tags in addition to finding those
+	// ASGs with spot-enabled=true
+	FilterByTags Tags
```
I guess the code could be tested despite its simplicity, just for the sake of it

tootedom replied:
Are there any tests for the Config at the moment, I'm not finding any, but perhaps I'm not looking in the correct place.

How do we currently test the Config object?

xlr-8 commented on core/main.go:
```diff
@@ -37,6 +39,12 @@ func Run(cfg *Config) {
+	if len(cfg.FilterByTags) == 0 {
+		cfg.FilterByTags = []Tag{{Key: "spot-enabled", Value: "true"}}
+	}
+}
```
This code could benefit of tests

tootedom replied:
>> I think that references what I said in cloudformation/stacks/AutoSpotting/template.json

 
As per my comment above.  
What does cloudformation do with the "Default" field?  Is it just documentation?

----

added tests for addDefaultFilter in main_test.go

tootedom commented on cloudformation/stacks/AutoSpotting/template.json:
```diff
@@ -71,6 +71,11 @@
+    "FilterByTags": {
+      "Default": "",
+      "Description": "Comma separated list of tag=value on which to filter the ASGs that autospotting considers.  By default 'spot-enabled=true'.  Example: 'spot-enabled=true,environment=dev'",
+      "Type": "String"
```
I'm not au fait with cloudformation, and I do not know where cloud formation uses the "Default", is it just documentation?  

Instead of adding a default, can we say in the decription:  If left blank, the code adds spot-enabled=true when nothing is specified (like it does now)?  

i.e. the is no default of Regions, but it defaults to us-east-1 for example in autospotting.go

I'll update the description, with this, and will await feedback

xlr-8 replied:
Fine with me!

tootedom commented on core/config.go:
```diff
@@ -1,12 +1,54 @@
+// String formats the ArrayOfStrings type string[]
+// into a string
+func (i *Tags) String() string {
+	if len(*i) > 0 {
```
I don't think this will ever be null, as the flags parsing is making it not.  All the examples I've seen, not one is checking for nil (see next comment for references),

And when running autospotting on command line, without the -tag_filters, it's not segmentation faulting

but I'll add it to be on the same side, if needed.

xlr-8 replied:
I know and I fully agree with you, there are indeed various pointers/slice pointers which aren't expected to be nil. I simply consider it a best practice to check anyway, as a basic reflex to avoid such case especially because it has little to no-cost performance wise.

tootedom commented on core/config.go:
```diff
@@ -1,12 +1,54 @@
+
+// Set supports the multiple command line options with the same
+// name
+func (i *Tags) Set(value string) error {
```
It's part of the interface for parsing flags.  Can't get rid of error,
This is how you parse multiple command line arguments into a slice, example:

- https://blog.rapid7.com/2016/08/04/build-a-simple-cli-tool-with-golang/
- https://lawlessguy.wordpress.com/2013/07/23/filling-a-slice-using-command-line-flags-in-go-golang/

xlr-8 replied:
Indeed, good point that was my mistake

tootedom commented on core/config.go:
```diff
@@ -1,12 +1,54 @@
+// name
+func (i *Tags) Set(value string) error {
+	if len(value) > 0 {
+		// Allow both space- and comma-separated values for the region list.
```
Yup, will update.

tootedom commented on core/mock_test.go:
```diff
@@ -134,6 +137,12 @@ func (m mockASG) UpdateAutoScalingGroup(*autoscaling.UpdateAutoScalingGroupInput
+
+func (m mockASG) DescribeAutoScalingGroupsPages(input *autoscaling.DescribeAutoScalingGroupsInput, function func(*autoscaling.DescribeAutoScalingGroupsOutput, bool) bool) error {
+	function(m.dasgo, true)
+	return nil
```
It's for mocking out the call to the aws api, so instead of actually calling AWS's DescribeAutoScalingGroupsPages, what you are doing is calling the mock'd objects' implementation of the DescribeAutoScalingGroupsPages implementation, which passes in the canned output data

This page gives a better justice description that I can.

- https://aws.amazon.com/blogs/developer/mocking-out-then-aws-sdk-for-go-for-unit-testing/

tootedom commented on core/region.go:
```diff
@@ -26,6 +26,8 @@ type region struct {
 	enabledASGs []autoScalingGroup
 	services    connections
 
+	tagsToFilterASGsBy []Tag
```
Will change.

tootedom commented on core/region.go:
```diff
@@ -261,72 +270,78 @@ func (r *region) requestSpotInstanceTypes() ([]string, error) {
-			{Name: aws.String("key"), Values: []*string{aws.String("spot-enabled")}},
-			{Name: aws.String("value"), Values: []*string{aws.String("true")}},
-		},
+func isASGWithMatchingTags(asg *autoscaling.Group, tagsToMatch []Tag) bool {
```
Will change

tootedom commented on core/region.go:
```diff
@@ -261,72 +270,78 @@ func (r *region) requestSpotInstanceTypes() ([]string, error) {
+	return matchedTags == len(tagsToMatch)
+}
+
+func (r *region) findMatchingASGsInPageOfResults(groups []*autoscaling.Group, tagsToMatch []Tag) []autoScalingGroup {
```
Will change

tootedom commented on core/region.go:
```diff
@@ -261,72 +270,78 @@ func (r *region) requestSpotInstanceTypes() ([]string, error) {
-	var asgs []*string
+func isASGWithMatchingTag(tagToMatch Tag, asgTags []*autoscaling.TagDescription) bool {
+	for _, asgTag := range asgTags {
+		if tagsMatch(*asgTag, tagToMatch) {
```
Have changed.  Let me know if better.

tootedom commented on core/region.go:
```diff
@@ -261,72 +270,78 @@ func (r *region) requestSpotInstanceTypes() ([]string, error) {
+
+	var asgs []autoScalingGroup
+
+	for _, group := range groups {
```
Moved into isASGWithMatchingTags(...) check

tootedom commented on core/region.go:
```diff
@@ -261,72 +270,78 @@ func (r *region) requestSpotInstanceTypes() ([]string, error) {
+		names[*slice] = slice
+	}
+	return names
+}
```
Sorry, forgot to remove from a previous refactor

tootedom commented on terraform/variables.tf:
```diff
@@ -39,6 +39,11 @@ variable "asg_regions_enabled" {
 
+variable "asg_tag_filters" {
+  description = "Tags to filter which ASGs autospotting considers"
+  default     = ""
```
updated the comment

tootedom commented:
Hi there,

I'm hopefully addressed the comments, but please let me know if not.

----

For the README documentation.  Will take me a bit more time.  I'll see what I can do.


thanks
/dom

tootedom commented:
I've update the START.md with some changes.  I think that covers the setting.  Let me know.
thanks
/dom

tootedom commented:
Hola,

Thinking about what I'm gonna do is convert the Tags type in Config back to a string, and move the type conversion into the Region setup. This will remove complexity and address a couple of the concerns @xlr-8  had.  I'll get you this, this evening, barring any acts of nature.

Just on another note, post Thurs this week, it'll be hard for me to get to a computer for a couple of weeks.  Accept my apologies upfront if there's any slow feedback after that date.

cristim commented:
Thanks @tootedom!

I hope we can merge it by then with no issue left, but we can also merge it as far as you can get by Thursday and then we clean up any remaining issues after the fact.

tootedom commented:
Ok, I've updated to make things simpler.  Got rid of the "Tags" and "Tag" from config, and used a simple string instance.  Region.go converts the comma separated string to slice of []Tag.  Added test in region_test.go to test various conversions on the comma separated string.  I think this answers the last of your comments @xlr-8 regarding the nil checks.

Let me know what you think, and if you need any other updates.

cheers
/dom

xlr-8 commented:
@tootedom thanks! I'll take a look this week for sure!

cristim commented:
Thanks for your contribution to the project
